### DISCUSSION TITLE
Navigation issues when using a route with arguments

### DISCUSSION THREAD

**Discussion by irconde** [Milad-Akarie/auto_route_library#1500] [Q&A]
I'm trying to define to possible initial routes conditioned to user authentication. Everything worked as expected until I defined a parameter for the corresponding widgets (isFirstScreen)

```dart
routerDelegate: _appRouter.delegate(
          initialRoutes: authenticated
              ? [CoursesPageWidget(isFirstScreen: true)]
              : [WelcomePageWidget(isFirstScreen: true)]),
``` 

Here is how I defined one of those two widgets

```dart
@RoutePage()
class CoursesPageWidget extends ConsumerStatefulWidget {
  final bool? isFirstScreen;
  const CoursesPageWidget({Key? key, this.isFirstScreen = false}) : super(key: key);

  @override
  ConsumerState<CoursesPageWidget> createState() => _CoursesPageWidgetState();
}
``` 

Every time I try to run my app this is the error I get

```
E/flutter (11337): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Expected [WelcomePageWidgetArgs],  found [WelcomePageWidgetArgs]
E/flutter (11337): #0      RouteData.argsAs (package:auto_route/src/route/route_data.dart:94:7)
E/flutter (11337): #1      new $AppRouter.<anonymous closure> (package:adapt_clicker/backend/Router/app_router.gr.dart:112:30)
E/flutter (11337): #2      RootStackRouter._pageBuilder (package:auto_route/src/router/controller/root_stack_router.dart:146:20)
E/flutter (11337): #3      StackRouter._addEntry (package:auto_route/src/router/controller/routing_controller.dart:1407:29)
E/flutter (11337): #4      StackRouter._pushAllGuarded (package:auto_route/src/router/controller/routing_controller.dart:1385:29)
E/flutter (11337): <asynchronous suspension>
``` 

auto_route version: ^6.0.5
auto_route_generator version: ^6.0.3

I would appreciate it if somebody could please provide with some guidance on how to solve the issue. Thanks in advance!

**Comment by Milad-Akarie** [comment]
@irconde Not sure what's going on here but this a perfect use case for an AutoRouteGuard why not use that?

**Reply by irconde** [reply]
Thanks for the prompt response. To be honest... I am learning on the go and I was unaware of that option. We'll give it a try.

It is kind of odd, though. Everything was working fine until we migrated to v.6. Anyways... I'll try the AutoRouteGuard. Thanks again! Appreciate it.

**Reply by Milad-Akarie** [reply]
Let me know if you need any help with guards

**Reply by irconde** [reply]
So... I was able to work on this issue some time this week. According to my tests, when I define a route with arguments, I can use it successfully anywhere in my code except for the main file. 

In other words.  I have this widget

`@RoutePage()
class ContactUsWidget extends ConsumerStatefulWidget {
  final bool? openFromDrawer;

  const ContactUsWidget({Key? key, this.openFromDrawer = false})
      : super(key: key);

  @override
  ConsumerState<ContactUsWidget> createState() => _ContactUsWidgetState();
}`

And I can do, for instance, this in.a drawer menu:

`if (widget.currentSelected == DrawerItems.contact) return;
 context.pushRoute( ContactUsWidget(openFromDrawer: true, ),);`

but, then, if I try to do this in my main file:

`
@override
  Widget build(BuildContext context) {
    return MaterialApp.router(
     ...
      theme: ThemeData(
        brightness: Brightness.light,
        inputDecorationTheme: FlutterFlowTheme.of(context).inputTheme(),
        appBarTheme: FlutterFlowTheme.of(context).appBarTheme(),
      ),
      themeMode: _themeMode,
      routeInformationProvider: _appRouter.routeInfoProvider(),
      routerDelegate: _appRouter.delegate(
          initialRoutes: <PageRouteInfo<dynamic>>[ContactUsWidget(openFromDrawer: true)]),
      routeInformationParser: _appRouter.defaultRouteParser(includePrefixMatches: true),
    );
  }
}
`
it just says that there's an error:

`[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Expected [ContactUsWidgetArgs],  found [ContactUsWidgetArgs]`

that points to the following function in the package: 
`T argsAs<T>({T Function()? orElse}) {
    final args = _match.args;
    if (args == null) {
      if (orElse == null) {
        final messages = [
          '${T.toString()} can not be null because the corresponding page has a required parameter'
        ];
        if (_match.autoFilled) {
          messages.add(
              '${_match.name} is an auto created ancestor of target route ${_match.flattened.last.name}');
          messages.add(
              'This usually happens when you try to navigate to a route that is inside of a nested-router\nbefore adding the nested-router to the stack first');
          messages.add(
              'try navigating to ${_match.flattened.map((e) => e.name).join(' -> ')}');
        }
        throw FlutterError('\n${messages.join('\n')}\n');
      } else {
        return orElse();
      }
    } else if (args is! T) {
      throw FlutterError(
          'Expected [${T.toString()}],  found [${args.runtimeType}]');
    } else {
      return args;
    }
  }`

I would really appreciate any workaround for this. I have spent a lot of hours trying to figure it how to make this work. Something, btw, that strangely worked for older versions.
